log-levels.md

```diff
@@ -0,0 +1,203 @@
+
+#### Examples
+
+`trace` level logging:
+
+- could include various additional information about a request, such as various diagnostics about created datastructures, the state of caches or similar, which are created in order to serve a request.
+- could include "begin operation" and "end operation" logs, 
+  - however please do consider using swift-distributed-tracing to instrument such "begin/end" events as tracing can be more efficient than logging which may need to create string representations of the durations and log levels etc.
+
+`debug` level logging:
+
+- may include a single log statement for opening a connection, or accepting a request etc,
+- it can include a _high level_ overview of what is going on in the library "started work, processing step X, finished work X, result code 200"
+
+### Log levels to avoid
+
+All these rules are only _general_ guidelines, and as such may have exceptions.Consider the following examples and rationale for why logging at high log levels by a library may not be desirable:
+
+It is generally _not acceptable_ for a service client (e.g. an http client) to log an `error` when a request has failed. End-users may be using the client to probe if an endpoint is even responsive or not, and a failure to respond may be _expected_ behavior. Logging errors would only confuse and pollute their logs.
+Instead, libraries should either `throw`, or return an `Error` value that users of the library will have enough knowladge about if they should log or ignore it.
+
+It is even less acceptable for a library to log any successful operations. This leads to flodding server side systems, especially 
+
+#### Examples (of things to avoid)
+
+Avoid using `info` or any higher log level for:
+
+- "normal operation" of the library, i.e. there is no need to log on info level "accepted a request" as this is the normal operation of a web service.
+
+Avoid using `error` or `warning`:
+
+- to report errors which the end-user of the library has the means of logging themselfes, e.g. if a database driver fails to fetch all rows of a query, it should not log an error or warning, but instead e.g. error the stream of values (or function, async function, or even the async sequence) that was representing the returned values.
+  - since the end-user is consuming these values, and has a mean of reporting (or swallowing) this error, the library should not log anything on their behalf.
+- never report as warnings which is merely an information, e.g. "weird header detected" may look like a good idea to log as a warning at first sight, however if the "weird header" is simply a misconfigured client (or just a "weird browser") you may be accidentally completely flooding an end-users logs with these "weird header" warnings (!)
```

Comment:
We might want to have an additional explanation as to why not to log a warning once, then fall back to debug/trace. (Must always check an atomic)

Reply:
I expanded it a bit

It may be tempting to implement a "log as warning only once" technique for per-request style situations which may be almost important enough to be a warning, but should not be logged repeatedly after all. Authors may think of smart techniques to log a warning only once per "weird header discovered" and later on log the same issue on a different, e.g. trace level... Such techniques result in confusing hard to debug logs, where developers of a system unaware of the stateful nature of the logging would be left confused when trying to reproduce the issue. 

- For example, if a developer spots such warning in a productiuon system, they may attempt to reproduce it thinking that it only happens in the production environment. However if the logging systems' log level choice is _sateful_ they may actually be successfully reproducing the issue but never seeing it manifest. For this, and related performance reasons (as implementing "only once per X" implies growing storage and per-request additional checking requirements), it is not recommended to apply this pattern.